### PR TITLE
ptycho 0.14: weeding out bugs and bundling new features

### PR DESCRIPTION
**Pull request by smribet**
This PR has new functionality for `phase` including: 
- compatibility with py4DSTEM `0.14`
- the option for real potentials 
- plotting updates including complex probes 
- standardization of regularization options across classes
- regularization along z direction for multislice
- overlap tomography for magnetic potentials 
- read/write functionality 

Updated tutorials [here](https://github.com/py4dstem/py4DSTEM_tutorials/tree/main/notebooks/version_0.14)

### CODE REVIEW

File: py4DSTEM/process/phase/iterative_multislice_ptychography.py [hunk 1]
```diff
@@ -1784,50 +2125,53 @@ def _visualize_last_iteration_figax(
...
 
         """
         figsize = kwargs.get("figsize", (8, 5))
         cmap = kwargs.get("cmap", "magma")
+        invert = kwargs.get("invert", False)
+        hue_start = kwargs.get("hue_start", 0)
         kwargs.pop("figsize", None)
         kwargs.pop("cmap", None)
+        kwargs.pop("invert", None)
+        kwargs.pop("hue_start", None)
```
Comment (sezelt) [inline py4DSTEM/process/phase/iterative_multislice_ptychography.py:2166]: Same comment about combining get and pop

File: py4DSTEM/process/phase/iterative_simultaneous_ptychography.py [hunk 2]
```diff
@@ -622,25 +685,61 @@ def preprocess(
...
+            vmin = kwargs.get("vmin", None)
+            vmax = kwargs.get("vmax", None)
+            hue_start = kwargs.get("hue_start", 0)
+            invert = kwargs.get("invert", False)
             kwargs.pop("figsize", None)
             kwargs.pop("cmap", None)
+            kwargs.pop("vmin", None)
+            kwargs.pop("vmax", None)
+            kwargs.pop("hue_start", None)
+            kwargs.pop("invert", None)
```
Comment (sezelt) [inline py4DSTEM/process/phase/iterative_simultaneous_ptychography.py:733]: Merge

File: py4DSTEM/process/phase/iterative_simultaneous_ptychography.py [hunk 3]
```diff
@@ -2761,44 +2962,68 @@ def _visualize_last_iteration_figax(
...
+        vmin_e = kwargs.get("vmin_e", min_e)
+        vmax_e = kwargs.get("vmax_e", max_e)
+        vmin_m = kwargs.get("vmin_m", min_m)
+        vmax_m = kwargs.get("vmax_m", max_m)
+
+        kwargs.pop("vmin_e", None)
+        kwargs.pop("vmax_e", None)
+        kwargs.pop("vmin_m", None)
+        kwargs.pop("vmax_m", None)
+
```
Comment (sezelt) [inline py4DSTEM/process/phase/iterative_simultaneous_ptychography.py:3026]: Merge

File: py4DSTEM/process/phase/iterative_overlap_magnetic_tomography.py [hunk 4]
```diff
@@ -0,0 +1,3070 @@
...
+            vmax = kwargs.get("vmax", None)
+            hue_start = kwargs.get("hue_start", 0)
+            invert = kwargs.get("invert", False)
+            kwargs.pop("figsize", None)
+            kwargs.pop("cmap", None)
+            kwargs.pop("vmin", None)
+            kwargs.pop("vmax", None)
+            kwargs.pop("hue_start", None)
+            kwargs.pop("invert", None)
+
```
Comment (sezelt) [inline py4DSTEM/process/phase/iterative_overlap_magnetic_tomography.py:801]: Merge

File: py4DSTEM/process/phase/iterative_overlap_magnetic_tomography.py [hunk 5]
```diff
@@ -0,0 +1,3070 @@
...
+        vmin_e = kwargs.get("vmin_e", 0.0)
+        vmax_e = kwargs.get("vmax_e", max_e)
+        vmin_m = kwargs.get("vmin_m", -max_m)
+        vmax_m = kwargs.get("vmax_m", max_m)
+
+        kwargs.pop("vmin_e", None)
+        kwargs.pop("vmax_e", None)
+        kwargs.pop("vmin_m", None)
+        kwargs.pop("vmax_m", None)
```
Comment (sezelt) [inline py4DSTEM/process/phase/iterative_overlap_magnetic_tomography.py:2768]: Same

File: py4DSTEM/process/phase/iterative_overlap_tomography.py [hunk 6]
```diff
@@ -618,19 +674,100 @@ def preprocess(
...
+            vmin = kwargs.get("vmin", None)
+            vmax = kwargs.get("vmax", None)
+            hue_start = kwargs.get("hue_start", 0)
+            invert = kwargs.get("invert", False)
             kwargs.pop("figsize", None)
             kwargs.pop("cmap", None)
+            kwargs.pop("vmin", None)
+            kwargs.pop("vmax", None)
+            kwargs.pop("hue_start", None)
+            kwargs.pop("invert", None)
```
Comment (sezelt) [inline py4DSTEM/process/phase/iterative_overlap_tomography.py:746]: you know

File: py4DSTEM/process/phase/iterative_overlap_tomography.py [hunk 7]
```diff
@@ -2075,24 +2339,33 @@ def _visualize_last_iteration(
...
+            min/max y indices
         """
         figsize = kwargs.get("figsize", (8, 5))
         cmap = kwargs.get("cmap", "magma")
+        invert = kwargs.get("invert", False)
+        hue_start = kwargs.get("hue_start", 0)
         kwargs.pop("figsize", None)
         kwargs.pop("cmap", None)
+        kwargs.pop("invert", None)
+        kwargs.pop("hue_start", None)
```
Comment (sezelt) [inline py4DSTEM/process/phase/iterative_overlap_tomography.py:2368]: 😇

**Review by sezelt — CHANGES_REQUESTED** [review]
Excellent work, team. 
I have essentially two categories of comment:
1. In many places you redundantly use `kwargs.get(key,default)` followed by `kwargs.pop(key,None)`, which can be accomplished by one call to `kwargs.pop(key,default)`.
2. Some very severe regularizations are turned on by default now: `'potential'` mode on the object and `symmetrize_probe`. Do you have examples of experimental data where these make a substantial improvement?  These will undoubtedly help for simulated data with no inelastic scattering and no aberrations beyond C1, but my suspicion is that for experimental data these are too aggressive and should default to off.

**Comment by sezelt** [comment]
I have been informed that `symmetrize_probe` is indeed off by default. It is turned __off__ after zero iterations, not on.

File: py4DSTEM/process/phase/iterative_base_class.py [hunk 8]
```diff
@@ -846,11 +836,15 @@ def _solve_for_center_of_mass_relative_rotation(
...
+
+                    self._rotation_div = rotation_div
+                    self._rotation_div_transpose = rotation_div_transpose
                 else:
                     # Minimize Curl
                     ind_min = xp.argmin(rotation_curl).item()
                     ind_trans_min = xp.argmin(rotation_curl_transpose).item()
-
+                    self._rotation_curl = rotation_curl
+                    self._rotation_curl_transpose = rotation_curl_transpose
```
[the remark this answers is not in the document]
Comment (smribet) [reply py4DSTEM/process/phase/iterative_base_class.py:847]: I wanted to save these variables for making figures

File: py4DSTEM/process/phase/iterative_dpc.py [hunk 9]
```diff
@@ -311,6 +488,63 @@ def _update(
...
+        Butterworth filter used for low/high-pass filtering.
+
+        Parameters
+        --------
+        current_object: np.ndarray
+            Current object estimate
+        q_lowpass: float
+            Cut-off frequency in A^-1 for low-pass butterworth filter
+        q_highpass: float
+            Cut-off frequency in A^-1 for high-pass butterworth filter
```
[the remark this answers is not in the document]
Comment (smribet) [reply py4DSTEM/process/phase/iterative_dpc.py:524]: The datacube has to have real space calibrations!

File: py4DSTEM/process/phase/utils.py [hunk 10]
```diff
@@ -1047,97 +1040,247 @@ def return_1D_profile(
...
+        )[crop_slice].imag
+
+    return array_nd
+
+
+def my_idst_I(array_nd, xp=np):
+    """1D rfft-based iDST-I"""
+    scaling = np.prod((np.array(array_nd.shape) + 1) * 2)
+    return my_dst_I(array_nd, xp=xp) / scaling
+
```
[the remark this answers is not in the document]
Comment (smribet) [reply py4DSTEM/process/phase/utils.py:1240]: updated!

**Review by sezelt — APPROVED** [review]

**Comment by smribet** [comment]
Thank you for the helpful feedback!

**Pull request merged**